OOP Library

Charadon opened the pull request:
This isn't currently done, and i'm more opening this as a way to go "Hey, this exists."

This pull request basically adds a library that abstracts Raylib into a more Vala-like experience. The reason I went down the library route instead of doing it directly through the vapi was because I quickly realized it would make the raylib vapi a massive 10,000+ line unmaintainable monstrosity. This also allows the ability to keep the more c-like API alongside the OOP API.

This also has an unattended consequence of making it work on any language that supports GObject Introspection.

Another thing I did was add Raylib as a subproject, so people could use this binding as a subproject for their own projects and have the full stack.

The Util namespace just contains functions that I made to help me out, and thought it'd be useful in general.

The code is also documented, so using `valadoc` will create pretty good documentation.

Now, I don't think i'll be able to get around to doing the 3D stuff, as I don't really do anything with 3D.

With a bunnymark score of around 62,000 before it dips below 60fps. I'd say it performs pretty well, as well!

Charadon committed:
Initial work on OOP

Charadon committed:
Added definitions for the color constants

Charadon committed:
Added a utility source code file.

Charadon committed:
Finished Color Custom Object

Charadon committed:
I don't get it

Charadon committed:
Fixed issue with re-using the namespace of Raylib

Charadon committed:
Added Music object

Charadon committed:
Added FileType namespace in Audio to make from_memory loads easier

Charadon committed:
Added some test resources, and added MasterVolume namespace

Charadon committed:
Changed master volume to be part of the window object

Charadon committed:
Added missing LoadSoundAlias function

Charadon committed:
Massive overhaul of the build system

Charadon committed:
Removed debug file

Charadon committed:
Added options to raylib subproject

Charadon committed:
Moved around some stuff in the raylib subproject, and added bunnymark to it for testing purposes.

Charadon committed:
Added missing function

Charadon committed:
Re-created bunnymark in vala

Charadon committed:
Added Keyboard and Gamepad objects

Charadon committed:
Added position property to Mouse object

Charadon committed:
Fixed typo where y was assigned to y, instead of value being assigned to y

Charadon committed:
Added screen width properties to Window object

Charadon committed:
Made bunnymark more OOP

Charadon committed:
Added a draw method that makes use of Vala's lambda feature

Charadon committed:
Changed bunnymark to make use of new Window.draw() method, and switched out some Raylib functions with RaylibOOP equivalents

lxmcf commented:
Hey!

This looks awesome and I agree going down the library path is best, main thing that prevented me from working on a VAPI for it was; as you said, how massive it would end up!

Let me know when you think everything 2D wise is sorted and I have no issues with merging this across :-)

Charadon committed:
Created bin2vala, a tool that allows you to easily embed binary files into your source code. For use with .from_memory loading

Charadon committed:
bin2vala: Switch from hexidecimal to normal decimal. Saves a few bytes on the source file size.

Charadon committed:
bin2vala: Removed unneeded if statement

Charadon committed:
bin2vala: Minor fix

Charadon committed:
bin2vala: Made tabbing better. (Why must I do this?)

Charadon commented:
Thanks for the encouragement! Also, someone on the Vala matrix/IRC chat said to submit the c-api raylib vapi to vala-extra-vapis repository (https://github.com/GNOME/vala-extra-vapis).

We should totally do that once I get it updated to 5.0.0

Charadon committed:
Added error handling to Window.

Charadon committed:
Fixed double free issue, and updated window.vala test to use new error handling

Charadon committed:
Seperated Input objects into their files.

Charadon committed:
Added enum for Keyboard Keys. Totally didn't use GPT to quickly write this...

Charadon committed:
Added missing BORDERLESS_WINDOWED_MODE in ConfigFlags

Charadon committed:
Changed some KeyboardKey parameters to int

Charadon committed:
Added todo list

Charadon committed:
Done with mouse object

Charadon committed:
Added Clipboard object

Charadon committed:
Began work on making raylib use GLib's logging system, but due to a bug, it's not working right now.

Charadon committed:
Made Raylib use GLib Log functions. And added minimum size properties to Window object

Charadon committed:
Added missing log levels

Charadon committed:
Added take_screenshot method to Window Object

lxmcf commented:
Wouldn't be a bad idea, might help it get out there more!

Sorry I haven't updated the VAPI for 5.0 yet, been away from Vala for a while and working in straight C, had completely slipped my mind!

Charadon commented:
> Wouldn't be a bad idea, might help it get out there more!
> 
> Sorry I haven't updated the VAPI for 5.0 yet, been away from Vala for a while and working in straight C, had completely slipped my mind!

Ha, no worries. I've actually been away from Vala quite a bit too. Been f'ing around with objects in C directly. Surprisingly works well.

I also think I gave you access to my repo, so if you want to update 5.0 yourself, you can just use my repo so we don't have to deal with merge stuff

Only changes i've really made to the direct binding vapi is adding some missing functions added from 5.0.0, and changing the parameters to use `int`s instead of `enum` types. Since enums in vala are not really the same kind of enum in C, and it causes issues.

Charadon committed:
Added docker compose file to speed up testing on a rolling release distro and a stable distro

Charadon committed:
Changed some parameter enums to int, and fixed a compile issue with older vala version.

Charadon commented:
Oh yeah, forgot to bring this up. But you fine with me removing rini? GLib already has a built-in object called `GLib.KeyFile` which also handles ini files.

https://valadoc.org/glib-2.0/GLib.KeyFile.html

lxmcf commented:
That should be fine :-)

Ngl I am not sure why I added it as I never actually tested whether it worked

Charadon committed:
Removed rini, as GLib has it's own ini file handler (GLib.KeyFile)

Charadon committed:
Gamepad.vala: Error message not contains name of controller.

Charadon committed:
Added a bleeding edge dockerfile

Charadon committed:
Added ccache and fastestmirror to dockerfiles

Charadon commented:
So, I added a docker compose file to make testing on 3 major types of distros (stable, rolling and bleeding edge) easier. I actually caught some compilation issues this way on Almalinux 8, so it should work on any distro made past 2018, and it compiles on bleeding edge (Fedora Rawhide), so no future incoming compatibility issues =)

Charadon committed:
See if base64 works in readme

Charadon committed:
It doesn't, but atleast I shrank the png size by 73%!

Charadon committed:
Added position property to Window Object

Charadon commented:
Hey I want a 2nd opinion on this. Should the cursor functions be under the Window object, or the Input.Mouse (Or even Input.Mouse.Cursor) object? I'm conflicted since it's a graphical thing rather than a input related thing.

Refering to these functions:

```c
void ShowCursor(void);
void HideCursor(void);
bool IsCursorHidden(void);
void EnableCursor(void);
void DisableCursor(void);
bool IsCursorOnScreen(void);
```

Charadon committed:
Added touch object.

lxmcf commented:
For those functions specifically I would say the window object would be better considering they are more directly tied to the windows functionality, might be a bit strange to have it out of the 'mouse' object but think it makes more sense logically.

Charadon committed:
raylib.vapi: Fixed confusing function name.

Charadon committed:
Input/Touch.vala: Added gestures to touch object.

Charadon committed:
Window.vala: Added get_time property

Charadon committed:
Window.vala: Added frame_time property

Charadon committed:
raylib.vapi: Fixed function name error during linking

Charadon committed:
Window.vala: Added Cursor object in Window Object

Charadon committed:
Removed the Core namespace, as it was kind of redundant

Charadon committed:
Reshuffled the file tree

Charadon committed:
New Color Object

Charadon committed:
raylib.vapi: Added the 5.0.0 automation functions

Charadon commented:
I'm also gonna skip the VR stuff. Because I'm poor, and will probably never have access to a VR headset to test this crap lol

Charadon committed:
Automation/EventList.vala: Began work on EventList object

Charadon committed:
Window.vala: Added dropped files stuff to Window Object

Charadon committed:
raylib.vapi: Fixed array length issue in FilePathList

Charadon committed:
Colors.vala: NOW the predefined colors work.

Charadon committed:
Window.vala: Fixed oversight where I unloaded FilePathList before it was used.

Charadon committed:
meson.build: Added dropped file test. It'll print the file path in the console.

Charadon committed:
tests/dropped.vala: Moved foreach  to try block

Charadon committed:
Added the Automation stuff

Charadon committed:
Fixed typo

Charadon committed:
Window.vala: Added scissors mode and a monitor array

Charadon committed:
Window.vala: Added event waiting functions

Charadon committed:
Window.vala: Added max height properties

Charadon committed:
Window.vala: Fixed typo

Charadon committed:
Window.vala: Fixed segfault

Charadon committed:
Removed test executables, and made the raylib subproject static

Charadon committed:
Fixed typo

Charadon committed:
Color.vala: Switched Color object to use new internal variable technique.

Charadon committed:
Shapes/Vector2.vala: Switched object to use new internal variable technique

Charadon committed:
Started on Text/Font object

Charadon committed:
Added core_basic_window example, and a game template.

lxmcf commented:
VAPI has been updated to match Raylib 5.0, changes shouldn't break anything.

Charadon commented:
So, if you've noticed a bit of a decrease in activity in my repo. Blame Helldivers lol

But I'll get it done! Eventually...

lxmcf commented:
110% understandable, the star wars battlefront collection is about to remove all free time I have so I feel that!

Charadon committed:
Update gitignore

Charadon committed:
Removed vscode files

Charadon committed:
Added some MacOS stuff to Raylib's meson.build.

Unknown if this lets it compile as I don't have a mac to test it on.

Charadon committed:
Added comment on how to use bin2vala

Charadon committed:
Added Version constant and added a newline to the license.

Charadon committed:
Changed Reverse DNS to be the git repo rather than iotib.net

Charadon committed:
Changed  the destroy window message to a debug message.

Charadon committed:
Changed some comments in the templata source file.

Charadon committed:
Fix warnings when using GCC 13

Vala seems to produce quite of a bit of warnings on GCC 13.

So I added some -Wno args to disable them to stop clogging the output.

Charadon committed:
Implemented as much of core_basic_screen_manager as possible.

Shapes.Rectangle needs to be made before I can continue.

Charadon committed:
Implemented Shapes.Rectangle object.

Charadon commented:
Aw yeah, back in action! Gonna start re-creating the examples from Raylib upstream as that'll be a good way to indicate what needs to be implemented.

As a sidenote, I love that Raylib is API stable, so I don't feel rushed at all to accomplish this =D

Like, 6.0 could come out tomorrow, and i'd probably have to do minimal effort to get it working with it.

Charadon committed:
Fixed compilation errors when compiling on Almalinux 8 (stable)

Charadon committed:
Wrong license in screen manager example.

Charadon committed:
Added 2DCamera object.

Also moved the 2D Draw functions out of Window and into 2DCamera

Charadon committed:
Added line object and example/test using it.

Charadon committed:
lib/Shapes/Line.vala: Fixed compilation issue on (stable)

Charadon committed:
Added free_functions to the Raylib vapi

Charadon committed:
Line.vala: Added new method draw_gl and changed draw to use Raylib.draw_line()

I didn't know this, but depending on the function used, raylib uses a different technique to draw it. This can cause unexpected behavior.

Charadon committed:
Clipboard.vala: Turn clipboard object in a non-glib object.

This is so I don't have to override get and set methods. Makes the code simpler.

Charadon committed:
Color.vala: Added the fade method.

Charadon committed:
Sound.vala: Updated internal variable to match other objects.

Charadon committed:
Rectangle.vala: Fix bug where I used the origin variable wrong.

Charadon committed:
Music.vala: Made the internal variable naming scheme match the rest of the program.

Charadon committed:
meson.build: Put the example stuff into it's own file.

Charadon committed:
Finished core_2d_camera example.

Charadon commented:
Found out if I want to make all the examples, I gotta add raymath to the vapi file, and then OOP it 😅

lxmcf commented:
Ah yes!

I was working on a raymath VAPI a while ago but couldn't work out an easy way to make it 'portable' and usable without the raylib VAPI itself

Charadon commented:
> Ah yes!
> 
> I was working on a raymath VAPI a while ago but couldn't work out an easy way to make it 'portable' and usable without the raylib VAPI itself

Interestingly, a lot of the float math functions are already methods in vala's float object (Such as `clamp`). I might ask around to see if I could somehow extend the float object to have the other functions such as `lerp`

Now as for adding those functions to the vapi, i'm gonna err on the side of simplicity and just add it into it's own namespace `Raylib.Math` inside the `raylib.vapi` file.

lw64 commented:
> changing the parameters to use ints instead of enum types. Since enums in vala are not really the same kind of enum in C, and it causes issues.

Hi, I was wondering what your specific issue here was?

Charadon commented:
> > changing the parameters to use ints instead of enum types. Since enums in vala are not really the same kind of enum in C, and it causes issues.
> 
> Hi, I was wondering what your specific issue here was?

If I remember correctly, I was getting type errors at compilation. However, I think it was only an issue on 0.48, the one that comes with RHEL 8

lw64 commented:
ah I see. If the errors are still there it might be fixable with attributes.

halfmexican committed:
add enums

halfmexican committed:
formatting

Charadon committed:
Merge pull request #1 from halfmexican/main

Add Enums to Gamepad Class

lxmcf commented:
Just a heads up I will be updating the base VAPI to raylib 5.5 in the coming days, will try not to touch anything else to avoid further conflicts! 😄

Charadon commented:
> Just a heads up I will be updating the base VAPI to raylib 5.5 in the coming days, will try not to touch anything else to avoid further conflicts! 😄

Raysan is pretty good about backwards compatibility, so it shouldn't break anything... I hope lol
